<div style="background:linear-gradient(135deg,#1e1b4b 0%,#4338ca 55%,#6366f1 100%);border-radius:18px;padding:34px 30px;color:#fff;font-family:Inter,Segoe UI,sans-serif">
  <div style="font-size:12px;letter-spacing:3px;color:#c7d2fe;font-weight:700;text-transform:uppercase">Chapter 36 · Solutions</div>
  <div style="font-size:34px;font-weight:900;line-height:1.1;margin:10px 0 6px">Simulation, Worked Answers ✅</div>
  <div style="font-size:15px;color:#e0e7ff;max-width:720px;line-height:1.6">Full solutions to the five challenges. Try them yourself first, then compare.</div>
  <div style="margin-top:16px;font-size:13px;color:#cbd5e1">Statistics, Data Science and AI: A Visual Handbook · John Fisher · 2026</div>
</div>

## ⚙️ Setup

In [1]:
import numpy as np
rng = np.random.default_rng(361)
print("ready")

ready


<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#4f46e5;letter-spacing:1px">CHALLENGE 1 · ESTIMATE A PROBABILITY</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">At least one head in 3 flips</div>
<div style="color:#4a5578;margin-top:6px">Simulate 200,000 sets of 3 fair coin flips and estimate P(at least one head). Compare to the exact value.</div>
</div>

In [2]:
N=200_000
flips = rng.integers(0,2,size=(N,3))     # 1 = heads
est = (flips.sum(axis=1) >= 1).mean()
exact = 1 - 0.5**3
print(f"estimate = {est:.4f}")
print(f"exact 1 - (1/2)^3 = {exact:.4f}")

estimate = 0.8746
exact 1 - (1/2)^3 = 0.8750


**Answer:** the estimate matches the exact 7/8 = **0.875**. "At least one" is the complement of "none", and P(no heads) = (1/2)&sup3; = 1/8, so the answer is 1 &minus; 1/8 = 7/8. Simulation confirms it without the algebra.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#2563eb;letter-spacing:1px">CHALLENGE 2 · ESTIMATE AN EXPECTATION</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">The max of two dice</div>
<div style="color:#4a5578;margin-top:6px">Estimate E[max of two dice] by simulation. (This is fiddly to do by hand, which is exactly when simulation shines.)</div>
</div>

In [3]:
N=500_000
d = rng.integers(1,7,size=(N,2))
est = d.max(axis=1).mean()
print(f"E[max of two dice] estimate = {est:.4f}")
print(f"exact 161/36 = {161/36:.4f}")

E[max of two dice] estimate = 4.4743
exact 161/36 = 4.4722


**Answer:** E[max] &approx; **4.47** (exactly 161/36 = 4.472). To estimate an expectation you average the values instead of counting hits, the same Monte Carlo loop with `.mean()` over the quantity itself.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#059669;letter-spacing:1px">CHALLENGE 3 · BIRTHDAY PROBLEM</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">A surprising collision</div>
<div style="color:#4a5578;margin-top:6px">In a room of 23 people, estimate the probability that at least two share a birthday. The famous answer surprises almost everyone.</div>
</div>

In [4]:
N=200_000; k=23
def has_match():
    bdays = rng.integers(0,365,k)
    return len(np.unique(bdays)) < k
est = np.mean([has_match() for _ in range(N//20)])
print(f"P(shared birthday, 23 people) estimate = {est:.3f}")

P(shared birthday, 23 people) estimate = 0.506


**Answer:** about **0.507**, just over half, with only 23 people. The count of *pairs* is C(23,2) = 253, which is what drives the collisions, not the 23 people. The result feels impossible until you simulate it.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#db2777;letter-spacing:1px">CHALLENGE 4 · MONTE CARLO INTEGRATION</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Estimate an integral</div>
<div style="color:#4a5578;margin-top:6px">Estimate the integral of x^2 from 0 to 1 by averaging f(U) for uniform random U on [0,1]. The exact value is 1/3.</div>
</div>

In [5]:
N=1_000_000
u = rng.random(N)
est = (u**2).mean()        # E[f(U)] approximates the integral over [0,1]
print(f"Monte Carlo integral estimate = {est:.4f}")
print(f"exact 1/3 = {1/3:.4f}")

Monte Carlo integral estimate = 0.3335
exact 1/3 = 0.3333


**Answer:** the estimate lands near **0.333**. For a uniform U on [0,1], E[f(U)] equals the integral of f over [0,1], so averaging f(U) over many samples approximates the integral. This is how Monte Carlo tackles integrals too hard to do analytically, the core trick behind a great deal of computational statistics.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#7c3aed;letter-spacing:1px">CHALLENGE 5 · THE BOOTSTRAP</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">A confidence interval with no formula</div>
<div style="color:#4a5578;margin-top:6px">Given a sample, estimate a 95% confidence interval for its median by resampling with replacement 10,000 times.</div>
</div>

In [6]:
sample = rng.normal(100, 20, size=50)
B=10_000
boot_meds = np.array([np.median(rng.choice(sample, size=len(sample), replace=True)) for _ in range(B)])
lo, hi = np.percentile(boot_meds, [2.5, 97.5])
print(f"sample median = {np.median(sample):.2f}")
print(f"bootstrap 95% CI for median = [{lo:.2f}, {hi:.2f}]")

sample median = 100.83
bootstrap 95% CI for median = [96.45, 111.22]


**Answer:** the bootstrap resamples the data thousands of times, computes the median of each resample, and reads the 2.5th and 97.5th percentiles as the interval. The median has no tidy standard-error formula, but Monte Carlo does not care, it gives a confidence interval for *any* statistic you can compute.

---
<div style="background:#ffffff;border:1px solid #e6e9f2;border-radius:16px;padding:22px 26px;font-family:Inter,sans-serif">
<div style="font-size:19px;font-weight:800;color:#1a2138">🎉 Nicely done</div>
<div style="color:#4a5578;line-height:1.8;margin-top:8px">You estimated probabilities and expectations, cracked the birthday problem, integrated by sampling, and bootstrapped a confidence interval for a statistic with no formula. Monte Carlo turns "I cannot derive this" into "I can simulate this", which is why it threads through the rest of the book.</div>
</div>

<div style="text-align:center;color:#8b94b3;font-size:12px;margin-top:16px">Statistics, Data Science and AI: A Visual Handbook · © 2026 John Fisher</div>